fms/utils/generation.py

```diff
@@ -51,13 +57,36 @@ def generate(
         num_beams: TODO: support beam search
         use_cache: requires that the model accept use_cache and
             past_key_value_states args in forward method.
+        pad_id: int
+            the pad token id to use in case of batch generation where sequences are of different lengths
     """
     batched = False
+    requires_padding = False
     if num_beams != 1:
         raise NotImplementedError("generate() does yet not support beam search")
-    if type(input_ids) == torch.Tensor:
+    if isinstance(input_ids, torch.Tensor):
+        # TODO: nested tensors
         if input_ids.dim() != 1:
             batched = True
+    elif isinstance(input_ids, list):
+        max_length = max(p.size(0) for p in input_ids)
+        requires_padding = any(p.size(0) < max_length for p in input_ids)
+        if requires_padding:
+            input_ids = [
+                torch.cat(
+                    (
+                        torch.tensor(
+                            [pad_id] * (max_length - p.size(0)),
+                            device=p.device,
+                            dtype=torch.long,
+                        ),
+                        p,
+                    )
+                )
+                for p in input_ids
+            ]
+        input_ids = torch.stack(input_ids, dim=0)
```

Comment:
I think there should be some kind of helper function to do the batching/padding

Reply:
if nestedtensor compiles with sdpa now though, can we basically just do `input_ids = torch.nested.nested_tensor( input_ids )` ? then I think we wouldn't need the mask either?

Reply:
As long as sdpa knows the input is a nested tensor, and is properly utilizing the context lengths of each sequence, we can do this.

Reply:
I'd give the option for now, as it might very well be some other operation is still not implemented for nested tensors in our code, which might make it all break

Reply:
this being said, I do want to have the option to test with nested tensors, as this is the perfect testbed for all the new code we are adding to pytorch core to support them

Reply:
let's at least give it a shot. I'd rather not have to maintain/support both paths if they aren't both necessary

Reply:
I'll try this with nested tensors and update here